src/main/scala/mesosphere/marathon/core/election/impl/CuratorElectionService.scala

```diff
@@ -0,0 +1,195 @@
+package mesosphere.marathon.core.election.impl
+
+import akka.actor.ActorSystem
+import akka.event.EventStream
+import com.codahale.metrics.MetricRegistry
+import mesosphere.chaos.http.HttpConf
+import mesosphere.marathon.MarathonConf
+import mesosphere.marathon.core.base.ShutdownHooks
+import mesosphere.marathon.metrics.Metrics
+import org.apache.curator.{ RetrySleeper, RetryPolicy }
+import org.apache.curator.framework.{ CuratorFramework, CuratorFrameworkFactory }
+import org.apache.curator.framework.recipes.leader.{ LeaderLatch, LeaderLatchListener }
+import org.apache.zookeeper.{ ZooDefs, KeeperException, CreateMode }
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.{ Await, Future }
+import scala.util.control.NonFatal
+import scala.concurrent.duration._
+
+class CuratorElectionService(
+  config: MarathonConf,
+  system: ActorSystem,
+  eventStream: EventStream,
+  http: HttpConf,
+  metrics: Metrics = new Metrics(new MetricRegistry),
+  hostPort: String,
+  backoff: ExponentialBackoff,
+  shutdownHooks: ShutdownHooks) extends ElectionServiceBase(
+  config, system, eventStream, metrics, backoff, shutdownHooks
+) {
+  private lazy val log = LoggerFactory.getLogger(getClass.getName)
+
+  private lazy val client = provideCuratorClient()
+  private var latch: Option[LeaderLatch] = None
+
+  override def leaderHostPortImpl: Option[String] = synchronized {
+    try {
+      latch match {
+        case None => None
+        case Some(l) =>
+          val participant = l.getLeader
+          if (participant.isLeader) Some(participant.getId) else None
+      }
+    }
+    catch {
+      case NonFatal(e) =>
+        log.error("error while getting current leader", e)
+        None
+    }
+  }
+
+  override def offerLeadershipImpl(): Unit = synchronized {
+    log.info("Using HA and therefore offering leadership")
+    latch match {
+      case Some(l) =>
+        log.error("Offering leadership while being candidate")
+        l.close()
+      case _ =>
+    }
+    latch = Some(new LeaderLatch(
+      client, config.zooKeeperLeaderPath + "-curator", hostPort, LeaderLatch.CloseMode.NOTIFY_LEADER
+    ))
+    latch.get.addListener(Listener) // idem-potent
```

Comment:
`maybeLatch.foreach(_.addListener(Listener))` is a bit more idiomatic.

We set latch to a new value in the line before, so the comment doesn't seem to be necessary.